pkg/server/plugin_invoker.go

```diff
@@ -64,48 +74,213 @@ type invokerError struct {
+
+	// Temp: assume Accept is *per Call()* and that an input needs to first be received
+	acceptC := make(chan []string, 1)
+
+	// Temp: assume it's the invoker responsibility to propagate the correlationId when present
+	// this will only make sense with f: X -> Y functions
+	correlationIdC := make(chan []string, 1)
+
+	// Sidecar => function input
+	go func() {
+		for {
+
+			in, err := callServer.Recv()
+			if err == io.EOF {
+				input.Close()
+				errs <- nil
+				Log.Printf("[S->F] Reached EOF\n")
+				break
+			}
+			if status.Code(err) == codes.Canceled {
+				Log.Printf("[S->F] Canceled by other goroutine? %#v\n", err)
+				input.Close()
+				errs <- err // no use
+				break
+			}
+			if err != nil {
+				Log.Printf("[S->F] Error returned from callServer.Recv: %#v\n", err)
+				input.Close()
+				errs <- err
+				break
+			}
+
+			if in.Headers[Accept] != nil {
+				select {
+				case acceptC <- in.Headers[Accept].Values:
+				default:
+				}
+			}
+			if in.Headers[CorrelationId] != nil {
+				select {
+				case correlationIdC <- in.Headers[CorrelationId].Values:
+				default:
+				}
+			}
+
+			unmarshalled, err := pi.messageToFunctionArgs(in)
+			if err != nil {
+				Log.Printf("[S->F] Sending %v to errors\n", err)
+				input.Close()
+				errs <- err
+				break
+			}
+			Log.Printf("[S->F] About to send %v to function\n", unmarshalled)
+
+			//select {
+			//	case input <- unmarshalled:
+			//	case <-done: // by virtue of being closed somewhere else
+			//    break
+			//}
+			cases := []reflect.SelectCase{
+				{Dir: reflect.SelectSend, Chan: input, Send: reflect.ValueOf(unmarshalled)},
+				{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(done)},
+			}
+			chosen, _, recvOK := reflect.Select(cases)
+			if chosen == 1 {
+				if recvOK {
+					panic("illegal state: should only fall in this case because done channel was closed")
+				}
+				input.Close()
+				errs <- nil
+				break
+			}
+		}
+		Log.Printf("[S->F] Returning from sidecar => function input goroutine\n")
+	}()
+
+	// Function output => sidecar
+	go func() {
+		var accept []string
+
+		cases := []reflect.SelectCase{
+			{Dir: reflect.SelectRecv, Chan: output},
+		}
+		open := 1
+		if len(channelValues) == 2 { // user function has a (<-chan error) second result
+			cases = append(cases, reflect.SelectCase{Dir: reflect.SelectRecv, Chan: channelValues[1]})
+			open++
+		}
+
+		for {
+
+			chosen, value, more := reflect.Select(cases)
+			switch chosen {
+			case 0: // output
+				Log.Printf("[F->S] Returning %v more=%v\n", value, more)
+
+				if !more {
+					errs <- nil
+					cases[chosen].Chan = reflect.ValueOf(nil)
+					open--
+					break
+				}
 
-// propagateHeaders modifies and returns out by adding to it headers from in that belong to the requestReplyHeaders list
-func propagateHeaders(in *function.Message, out *function.Message) *function.Message {
-	for _, h := range requestReplyHeaders {
-		out.Headers[h] = in.Headers[h]
+				if accept == nil {
```

Comment:
yes, and this is OK IMO because ultimately:
Accept should be *per Call() invocation* and be a property of the downstream topic, taken care of by the sidecar (with an exception when said topic is _replies_, in which case streamsize==1)